bus-mapping/src/circuit_input_builder.rs

```diff
@@ -577,62 +645,205 @@ impl<'a> CircuitInputStateRef<'a> {
+                (CodeSource::Address(code_address), account.code_hash)
+            }
+        };
+
+        let get_memory_offset_length =
+            |nth: usize| -> std::result::Result<_, Error> {
+                let offset = step.stack.nth_last(nth)?;
+                let length = step.stack.nth_last(nth + 1)?;
+                if length.is_zero() {
+                    return Ok((0, 0));
+                }
+                Ok((offset.low_u64(), length.low_u64()))
+            };
+        let (
+            call_data_offset,
+            call_data_length,
+            return_data_offset,
+            return_data_length,
+        ) = match kind {
+            CallKind::Call | CallKind::CallCode => {
+                let call_data = get_memory_offset_length(3)?;
+                let return_data = get_memory_offset_length(5)?;
+                (call_data.0, call_data.1, return_data.0, return_data.1)
+            }
+            CallKind::DelegateCall | CallKind::StaticCall => {
+                let call_data = get_memory_offset_length(2)?;
+                let return_data = get_memory_offset_length(4)?;
+                (call_data.0, call_data.1, return_data.0, return_data.1)
             }
+            CallKind::Create | CallKind::Create2 => (0, 0, 0, 0),
         };
-        self.push_call(kind, address, code_source, code_hash);
+
+        let caller = &self.tx.calls[self.tx_ctx.call_index()];
+        let call = Call {
+            call_id: self.block_ctx.rwc.0,
+            kind,
+            is_static: kind == CallKind::StaticCall || caller.is_static,
+            is_root: false,
+            is_persistent: caller.is_persistent && is_success,
+            is_success,
+            rw_counter_end_of_reversion: 0,
+            caller_address,
+            address,
+            code_source,
+            code_hash,
+            depth: caller.depth + 1,
+            value,
+            call_data_offset,
+            call_data_length,
+            return_data_offset,
+            return_data_length,
+        };
+        self.push_call(call);
+
         Ok(())
     }
 
     /// Handle a return step caused by any opcode that causes a return to the
     /// previous call context.
-    fn handle_return(
-        &mut self,
-        step: &GethExecStep,
-        next_step: &GethExecStep,
-    ) -> Result<(), Error> {
-        if self.tx_ctx.call_stack.len() == 1 {
-            return Err(Error::InvalidGethExecStep(
-                "handle_tx: call stack will be empty",
-                Box::new(step.clone()),
-            ));
-        }
-        let (_, call_ctx) = self
-            .tx_ctx
-            .pop_call_index_ctx()
-            .expect("call stack is empty");
-        // If the return was successful, accumulate the swc from the
-        // subcall.
-        if !next_step.stack.last()?.is_zero() {
-            self.tx_ctx.call_ctx_mut().swc += call_ctx.swc;
+    fn handle_return(&mut self, is_success: bool) -> Result<(), Error> {
```

Comment:
Yes! For 3. I would like to say "apply" reversion operations instead of "generate", since they are already generated and staged somewhere but not applied.